Add utility for bundling sample inputs with models

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35631 Add utility for bundling sample inputs with models**
* #35630 Expose JIT Module's "register_attribute" to Python

Summary:
Bundling sample inputs with our models with a standardized interface
will make it possible to write benchmarking and code-coverage tools that
call all models in a uniform way.  The intent is to make this a standard
for mobile models within Facebook.  Putting it in torch/utils so tests
can run on GitHub and because it might be useful for others as well.

`augment_model_with_bundled_inputs` is the primary entry point.  See
its docstring for usage information and the test for some example uses.

One design question I had was how much power should be available for
automatic deflating and inflating of inputs.  The current scheme gives
some automatic handling and a reasonable escape hatch
("_bundled_input_inflate_format") for top-level tensor arguments, but no
automatic support for (e.g.) tensors in tuples or long strings.  For
more complex cases, we have the ultimate escape hatch of just defining
_generate_bundled_inputs in the model.

Another design question was whether to add the inputs to the model or
wrap the model in a wrapper module that had these methods and delegated
calls to `forward`.  Because models can have other exposed methods and
attributes, the wrapped seemed too onerous.

Test Plan:
Unit test.

Differential Revision: [D20925013](https://our.internmc.facebook.com/intern/diff/D20925013)

dreiss committed:
Add utility for bundling sample inputs with models

Summary:
Bundling sample inputs with our models with a standardized interface
will make it possible to write benchmarking and code-coverage tools that
call all models in a uniform way.  The intent is to make this a standard
for mobile models within Facebook.  Putting it in torch/utils so tests
can run on GitHub and because it might be useful for others as well.

`augment_model_with_bundled_inputs` is the primary entry point.  See
its docstring for usage information and the test for some example uses.

One design question I had was how much power should be available for
automatic deflating and inflating of inputs.  The current scheme gives
some automatic handling and a reasonable escape hatch
("_bundled_input_inflate_format") for top-level tensor arguments, but no
automatic support for (e.g.) tensors in tuples or long strings.  For
more complex cases, we have the ultimate escape hatch of just defining
_generate_bundled_inputs in the model.

Another design question was whether to add the inputs to the model or
wrap the model in a wrapper module that had these methods and delegated
calls to `forward`.  Because models can have other exposed methods and
attributes, the wrapped seemed too onerous.

Test Plan:
Unit test.

dzhulgakov commented on torch/utils/bundled_inputs.py:
```diff
@@ -0,0 +1,148 @@
+    if not isinstance(model, torch.jit.ScriptModule):
+        raise Exception("Only ScriptModule is supported.")
+
+    forward_arg_types = [node.type() for node in list(model.graph.param_node().outputs())[1:]]
```
I'm not an expert - but another way would be iterating over model's schema. Probably they would give the same results

dreiss replied:
What is the model's schema and how do I iterate over it?

dzhulgakov replied:
```
m = torch.jit.script(torch.nn.Linear(10,10))
for arg in m.forward.schema.arguments:
  print(arg.name, arg.type)

> self __torch__.torch.nn.modules.linear.Linear
> input Tensor
```

dzhulgakov commented on torch/utils/bundled_inputs.py:
```diff
@@ -0,0 +1,148 @@
+        raise Exception("Only ScriptModule is supported.")
+
+    forward_arg_types = [node.type() for node in list(model.graph.param_node().outputs())[1:]]
+    deflated_inputs_type = torch._C.ListType(torch._C.TupleType(forward_arg_types))
```
@suo - is there a way to maybe use python types here and do some parsing?

dzhulgakov commented on torch/utils/bundled_inputs.py:
```diff
@@ -0,0 +1,148 @@
+        # Allow custom inflation expressions any tensor.
+        # For example, calling custom image-decoding ops.
+        # Or just use "{}" as the format string to ignore size limits.
+        if hasattr(arg, "_bundled_input_inflate_format"):
```
wait, you're suggesting adding attribute to a tensor? it's a bit hacky. Maybe allow input argument to be a callable and if so - invoke it?

dreiss replied:
Yes, it is a bit hacky.  Assuming anything callable is meant to be called seems error-prone.  If you want something cleaner, I think I'd want to go with a NamedTuple defined in this module that has two fields: deflated value and inflation format string.  Then we can just do an isinstance check.

dzhulgakov replied:
yeah, named tuple is probably better, current one is just too magical

zdevito commented on torch/utils/bundled_inputs.py:
```diff
@@ -0,0 +1,148 @@
+            return all_inputs
+        """))
+
+    # Define some helper methods.
```
My impression is that here is more code generation here than necessary. Couldn't this functionality be implemented as a normal TorchScript model, and then converted use torch.jit.script?

dreiss replied:
I don't see how, but maybe I'm missing something.  I want this to be available as a utility that can be run after training, regardless of what the top-level module is.  Since I don't want to wrap the top-level module (see commit message), this was the only way I thought of to add these helpers.

dreiss committed:
Update on "Add utility for bundling sample inputs with models"


Summary:
Bundling sample inputs with our models with a standardized interface
will make it possible to write benchmarking and code-coverage tools that
call all models in a uniform way.  The intent is to make this a standard
for mobile models within Facebook.  Putting it in torch/utils so tests
can run on GitHub and because it might be useful for others as well.

`augment_model_with_bundled_inputs` is the primary entry point.  See
its docstring for usage information and the test for some example uses.

One design question I had was how much power should be available for
automatic deflating and inflating of inputs.  The current scheme gives
some automatic handling and a reasonable escape hatch
("_bundled_input_inflate_format") for top-level tensor arguments, but no
automatic support for (e.g.) tensors in tuples or long strings.  For
more complex cases, we have the ultimate escape hatch of just defining
_generate_bundled_inputs in the model.

Another design question was whether to add the inputs to the model or
wrap the model in a wrapper module that had these methods and delegated
calls to `forward`.  Because models can have other exposed methods and
attributes, the wrapped seemed too onerous.

Test Plan:
Unit test.

dreiss commented:
(update is just lint)

dreiss committed:
Update on "Add utility for bundling sample inputs with models"


Summary:
Bundling sample inputs with our models with a standardized interface
will make it possible to write benchmarking and code-coverage tools that
call all models in a uniform way.  The intent is to make this a standard
for mobile models within Facebook.  Putting it in torch/utils so tests
can run on GitHub and because it might be useful for others as well.

`augment_model_with_bundled_inputs` is the primary entry point.  See
its docstring for usage information and the test for some example uses.

One design question I had was how much power should be available for
automatic deflating and inflating of inputs.  The current scheme gives
some automatic handling and a reasonable escape hatch
("_bundled_input_inflate_format") for top-level tensor arguments, but no
automatic support for (e.g.) tensors in tuples or long strings.  For
more complex cases, we have the ultimate escape hatch of just defining
_generate_bundled_inputs in the model.

Another design question was whether to add the inputs to the model or
wrap the model in a wrapper module that had these methods and delegated
calls to `forward`.  Because models can have other exposed methods and
attributes, the wrapped seemed too onerous.

Test Plan:
Unit test.

[ghstack-poisoned]

dreiss committed:
Update on "Add utility for bundling sample inputs with models"


Summary:
Bundling sample inputs with our models with a standardized interface
will make it possible to write benchmarking and code-coverage tools that
call all models in a uniform way.  The intent is to make this a standard
for mobile models within Facebook.  Putting it in torch/utils so tests
can run on GitHub and because it might be useful for others as well.

`augment_model_with_bundled_inputs` is the primary entry point.  See
its docstring for usage information and the test for some example uses.

One design question I had was how much power should be available for
automatic deflating and inflating of inputs.  The current scheme gives
some automatic handling and a reasonable escape hatch
("_bundled_input_inflate_format") for top-level tensor arguments, but no
automatic support for (e.g.) tensors in tuples or long strings.  For
more complex cases, we have the ultimate escape hatch of just defining
_generate_bundled_inputs in the model.

Another design question was whether to add the inputs to the model or
wrap the model in a wrapper module that had these methods and delegated
calls to `forward`.  Because models can have other exposed methods and
attributes, the wrapped seemed too onerous.

Test Plan:
Unit test.
